lib/FrontendTool/FrontendTool.cpp

```diff
@@ -219,38 +239,79 @@ static bool emitLoadedModuleTraceIfNeeded(ASTContext &ctxt,
+  for (auto &module : ctxt.LoadedModules) {
+    ModuleDecl *loadedDecl = module.second;
+    assert(loadedDecl && "Expected loaded module to be non-null.");
+    pathToModuleDecl.insert(
+      std::pair<StringRef, ModuleDecl *>(loadedDecl->getModuleFilename(),
+                                         loadedDecl));
+  }
 
-  // Canonicalise all the paths by opening them.
-  for (auto &dep : depTracker->getDependencies()) {
-    llvm::SmallString<256> buffer;
-    StringRef realPath;
-    int FD;
+  std::vector<SwiftModuleTraceInfo> swiftModules;
+  llvm::SmallString<256> buffer;
+  for (auto &depPath : depTracker->getDependencies()) {
+    StringRef realDepPath;
     // FIXME: appropriate error handling
-    if (llvm::sys::fs::openFileForRead(dep, FD, llvm::sys::fs::OF_None,
-                                       &buffer)) {
-      // Couldn't open the file now, so let's just assume the old path was
+    if (llvm::sys::fs::real_path(depPath, buffer,/*expand_tilde=*/true))
+      // Couldn't find the canonical path, so let's just assume the old one was
       // canonical (enough).
-      realPath = dep;
-    } else {
-      realPath = buffer.str();
-      // Not much we can do about failing to close.
-      (void)close(FD);
-    }
+      realDepPath = depPath;
+    else
+      realDepPath = buffer.str();
 
     // Decide if this is a swiftmodule based on the extension of the raw
     // dependency path, as the true file may have a different one.
-    auto ext = llvm::sys::path::extension(dep);
-    if (file_types::lookupTypeForExtension(ext) ==
-          file_types::TY_SwiftModuleFile) {
-      swiftModules.push_back(realPath);
+    // For example, this might happen when the canonicalized path points to
+    // a Content Addressed Storage (CAS) location.
+    auto moduleFileType =
+      file_types::lookupTypeForExtension(llvm::sys::path::extension(depPath));
+    if (moduleFileType == file_types::TY_SwiftModuleFile
+        || moduleFileType == file_types::TY_SwiftParseableInterfaceFile) {
+      auto dep = pathToModuleDecl.find(depPath);
+      assert(dep != pathToModuleDecl.end() && "Dependency must've been loaded.");
+      ModuleDecl* depMod = dep->second;
+      swiftModules.push_back({
+        /*modulePath=*/
+        realDepPath,
+        // TODO: There is an edge case which is not handled here.
+        // When we build a framework using -import-underlying-module, or an
+        // app/test using -import-objc-header, we should look at the direct
+        // imports of the bridging modules (a clang module in the first case and
+        // a swift module in the second), and mark those as our direct imports.
+        /*isImportedDirectly=*/
+        importedModulePaths.find(depPath) != importedModulePaths.end(),
+        /*isCompiledWithLibraryEvolution=*/
+        depMod->isResilient()
+      });
     }
   }
 
+  // Almost a re-implementation of reversePathSortedFilenames :(.
+  std::sort(
```

Comment:
If there's a better name for this, it might be better to pull it out into a `static` function, just to make the call site simpler.

Reply:
I don't think the same function is going to be used elsewhere (unless I create a more generic version which takes a projection function as an argument), so I left it inline. Kinda' like how John Carmack writes [here](http://number-none.com/blow/john_carmack_on_inlined_code.html).